deploy/runtools/run_farm.py

```diff
@@ -1,361 +1,424 @@
-        self._next_port += 1
+        self._next_switch_port += 1
         return retport
 
-class F1_Instance(EC2Inst):
-    FPGA_SLOTS: int = 0
-    fpga_slots: List[FireSimServerNode]
-
-    def __init__(self) -> None:
-        super().__init__()
-        self.fpga_slots = []
-
-    def get_num_fpga_slots_max(self) -> int:
-        """ Get the number of fpga slots. """
-        return self.FPGA_SLOTS
-
     def add_simulation(self, firesimservernode: FireSimServerNode) -> None:
         """ Add a simulation to the next available slot. """
-        assert len(self.fpga_slots) < self.FPGA_SLOTS
-        self.fpga_slots.append(firesimservernode)
+        assert len(self.sim_slots) < self.MAX_SIM_SLOTS_ALLOWED
+        self.sim_slots.append(firesimservernode)
         firesimservernode.assign_host_instance(self)
 
-class F1_16(F1_Instance):
-    instance_counter: int = 0
-    FPGA_SLOTS: int = 8
-
-    def __init__(self) -> None:
-        super().__init__()
-        self.instance_id = F1_16.instance_counter
-        F1_16.instance_counter += 1
-
-class F1_4(F1_Instance):
-    instance_counter: int = 0
-    FPGA_SLOTS: int = 2
-
-    def __init__(self) -> None:
-        super().__init__()
-        self.instance_id = F1_4.instance_counter
-        F1_4.instance_counter += 1
+class RunFarm(metaclass=abc.ABCMeta):
+    """Abstract class to represent how to manage run farm hosts (similar to `BuildFarm`).
+    In addition to having to implement how to spawn/terminate nodes, the child classes must
+    implement `mapper_*` functions to help topologies map run farm hosts (`Inst`s) to `FireSimNodes`.
+    """
 
-class F1_2(F1_Instance):
-    instance_counter: int = 0
-    FPGA_SLOTS: int = 1
+    def __init__(self, args: Dict[str, Any]) -> None:
+        self.args = args
 
-    def __init__(self) -> None:
-        super().__init__()
-        self.instance_id = F1_2.instance_counter
-        F1_2.instance_counter += 1
+    @abc.abstractmethod
+    def get_smallest_sim_host_handle(self, num_sims: int) -> str:
+        """Return the smallest run host type that supports greater than or
+        equal to num_sims simulations AND has available run hosts of that type
+        (according to run host counts you've specified in config_run_farm.ini).
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def allocate_sim_host(self, sim_host_handle: str) -> Inst:
+        """Let user allocate and use an run host (assign sims, etc.).
+        This deliberately exposes sim_host_handles to users, so that if
+        they know exactly how to map to a particular platform, they always
+        have an escape hatch."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_default_switch_host_handle(self) -> str:
+        """Get the default host instance type name that can host switch
+        simulations. """
+        raise NotImplementedError
```

Comment:
These functions are used by topologies to map `FireSimServerNodes` (and the like) to specific run farm hosts. Do we like this naming? Does this naming make sense?

Reply:
Right now this API works with "handles" which can be anything. In the AWS EC2 case, this "handle" is a name of an inst like `f1.2xlarge` while in the externally provisioned case it is a hostname. Do we like "handle" as the name?

Reply:
I updated the docstrings. LMK if they need more clarification.